components/server/src/util/request-context.ts

```diff
@@ -7,32 +7,98 @@
-    contextTimeMs?: number;
+    /**
+     * Unique, artificial ID for this request.
+     */
+    readonly requestId: string;
+
+    /**
+     * A request kind e.g. "job", "http" or "grpc".
+     */
+    readonly requestKind: string;
+
+    /**
+     * A name. Specific values depends on requestKind.
+     */
+    readonly requestMethod: string;
+
+    /**
+     * Propagate cancellation through the handler chain.
+     */
+    readonly signal: AbortSignal;
+
+    /**
+     * The UNIX timestamp in milliseconds when request processing started.
+     */
+    readonly startTime: number;
+
+    /**
+     * The SubjectId this request is authenticated with.
+     */
+    readonly subjectId?: SubjectId;
+
+    /**
+     * The trace ID for this request. This is used to correlate log messages and other events.
+     */
+    readonly traceId?: string;
 }
 
 const asyncLocalStorage = new AsyncLocalStorage<RequestContext>();
+
+export function ctx(): RequestContext {
+    const ctx = asyncLocalStorage.getStore();
+    if (!ctx) {
+        throw new Error("getRequestContext: No request context available");
+    }
+    return ctx;
+}
+
+export function tryCtx(): RequestContext | undefined {
+    return asyncLocalStorage.getStore() || undefined;
+}
+
 /**
- * !!! Only to be used by selected internal code !!!
+ * @deprecated Only used during the rollout period. Use `getSubjectId` instead
+ */
+export function tryGetSubjectId(): SubjectId | undefined {
+    const ctx = tryCtx();
+    return ctx?.subjectId;
+}
+
+/**
+ * The context all our request-handling code should run in.
+ * By default, all fields are inhereted from the parent context. Only exceptions: `requestId` and `contextId`.
+ * @param subjectId If this undefined, the request is considered unauthorized
+ * @param contextKind
+ * @param context
+ * @param fun
  * @returns
  */
-export function getGlobalContext(): RequestContext | undefined {
-    return asyncLocalStorage.getStore();
+export function runWithRequestContext<T>(
+    context: Omit<RequestContext, "requestId" | "startTime"> & { requestId?: string; startTime?: number },
+    fun: () => T,
+): T {
+    const requestId = context.requestId || v4();
+    const startTime = context.startTime || performance.now();
+    return runWithContext({ ...context, requestId, startTime }, fun);
 }
 
-export function runWithContext<C extends RequestContext, T>(contextKind: string, context: C, fun: () => T): T {
-    return asyncLocalStorage.run(
-        {
-            ...context,
-            contextKind,
-            contextId: context.contextId || v4(),
-            contextTimeMs: context.contextTimeMs || performance.now(),
-        },
-        fun,
-    );
+export function runWithChildContext<T>(child: Pick<RequestContext, "subjectId">, fun: () => T): T {
+    const parent = ctx();
+    if (!parent) {
+        throw new Error("runWithChildContext: No parent context available");
+    }
+    // TODO(gpl) Here we'll want to create a new spanID for tracing
```

Comment:
I think we rather want to create a new childContext when we create a span (i.e. the other way around).